core/src/callback.jl

```diff
@@ -393,22 +398,131 @@ function set_fractional_flow_in_allocation!(
     return nothing
 end
 
-function set_control_params!(p::Parameters, node_id::NodeID, control_state::String)
-    node = getfield(p, p.graph[node_id].type)
-    idx = searchsortedfirst(node.node_id, node_id)
-    new_state = node.control_mapping[(node_id, control_state)]
+function update_parameters!(pump::Pump, parameter_update::ParameterUpdate)::Nothing
+    if parameter_update.node_type != NodeType.Pump
+        return nothing
+    end
+    (; node_idx, active, flow_rate_scalar) = parameter_update
+    pump.active[node_idx] = active
+    if !isnan(flow_rate_scalar)
+        get_tmp(pump.flow_rate, 0)[node_idx] = clamp(
+            flow_rate_scalar,
+            pump.min_flow_rate[node_idx],
+            pump.max_flow_rate[node_idx],
+        )
+    end
+    return nothing
+end
 
-    for (field, value) in zip(keys(new_state), new_state)
-        if !ismissing(value)
-            vec = get_tmp(getfield(node, field), 0)
-            vec[idx] = value
-        end
+function update_parameters!(outlet::Outlet, parameter_update::ParameterUpdate)::Nothing
+    if parameter_update.node_type != NodeType.Outlet
+        return nothing
+    end
+    (; node_idx, active, flow_rate_scalar) = parameter_update
+    outlet.active[node_idx] = active
+    if !isnan(flow_rate_scalar)
+        get_tmp(outlet.flow_rate, 0)[node_idx] = clamp(
+            flow_rate_scalar,
+            outlet.min_flow_rate[node_idx],
+            outlet.max_flow_rate[node_idx],
+        )
+    end
+    return nothing
+end
 
-        # Set new fractional flow fractions in allocation problem
-        if is_active(p.allocation) && node isa FractionalFlow && field == :fraction
```

Comment:
So this was clearly a red flag in the old code, although I like the genericity of the rest.